### PR TITLE
Component scaffolding generator

### LINKED ISSUE #1323: Component scaffolding generator
## Problem

When creating a new component, there's a few things you need to add in various places that can make the initial set up a little cumbersome:

1. The SFC itself
2. The tsconfig.json includes entry
3. The storybook `.mdx` file (with all it's boilerplate)
4. The unit test file
5. The storybook snapshot tests

All of this could be templated out using something like lodash templates (or maybe something else :shrug:) and a package.json script could be created to scaffold out all the support for a new component.

## Description

To keep things simple, we could use something like https://github.com/ItsJonQ/remake to implement the whole process top-to-bottom.

Adding the `tsconfig.json` includes entry would need to happen separate from remake as I don't believe it has the ability to update existing files, just generate new ones.

## Implementation

- [ ] 🙋 I would be interested in implementing this feature.

### PR DESCRIPTION
**Pull request by anton202**
## Fixes

Fixes #1323 by @sarayourfriend 

## Description

This PR adds the ability to crate a component scaffolding from the command line.
by runing: `pnpm run create:component [component name in PascalCase]`
this files will be created:
- the component file
- story file
- unit test file
- regression test file
- the componet name will be added to tsconfig.ts includes property

## Testing Instructions

you'll find the details in the main readme file under Frontend Components --> How to create a component,  on what commands to run and how to run them in order to create the component scafolding

also please take a look at the template files inside .remake folder, to see if i missed something in the templates

## Checklist

- [x] My pull request has a descriptive title (not a vague title like `Update index.md`).
- [x] My pull request targets the *default* branch of the repository (`main`) or a parent feature branch.
- [x] My commit messages follow [best practices][best_practices].
- [x] My code follows the established code style of the repository.
- [ ] I added or updated tests for the changes I made (if applicable).
- [x] I added or updated documentation (if applicable).
- [x] I tried running the project locally and verified that there are no visible errors.

[best_practices]:https://git-scm.com/book/en/v2/Distributed-Git-Contributing-to-a-Project#_commit_guidelines

## Developer Certificate of Origin

<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>

### CODE REVIEW

File: tsconfig.json
Comment (sarayourfriend) [inline tsconfig.json:97]: Do you know where this came from?

**Review by sarayourfriend — CHANGES_REQUESTED** [review]
LGTM so far! Just a few changes to the documentation necessary as we use `pnpm`, not plain `npm`.

File: tsconfig.json
Comment (anton202) [reply tsconfig.json:97]: yeah :), comment-json removes all white lines from tsconfig file after stringifing it back to json.
i add this line as a seperator(before there was a white line at line 97) between the section of non components and the components that will be add in the future right after this section, if one uses the command line to create the component.
i can remove it if you``ll think its not necessary.

Comment (sarayourfriend) [reply tsconfig.json:97]: Ah, I see! It might be better to replace the string of `*` with actual text, otherwise it looks like something prime for deletion by someone with less context down the road :slightly_smiling_face:

**Review by sarayourfriend — COMMENTED** [review]
@anton202 When you `pnpm install` locally, do you get an error about a missing minimist version (1.2.5)? It doesn't look like that version is present in the lockfile anymore but was previously. I don't see any packages referencing it either. But when I run `pnpm install` I get this:

```
Lockfile is up to date, resolution step is skipped
 WARN  Broken lockfile: no entry for '/minimist/1.2.5' in pnpm-lock.yaml
 ERR_PNPM_ERR_PNPM_LOCKFILE_MISSING_DEPENDENCY  The lockfile is broken! Resolution step will be performed to fix it.
```

```
pnpm --version
7.17.0
```

Any ideas?

**Comment by anton202** [comment]
i think i forgot to commit the changes in the lock file when i merged the main branch

File: README.md
Comment (obulat) [inline README.md:130]:
```suggestion
This command will create a component file, a story file for the component, a unit test file, and a regression test file. It also adds the component name to tsconfig.ts `includes` property.
```

File: README.md
Comment (obulat) [inline README.md:131]:
```suggestion
Each file will have a basic template to start working with. We use [itsjonq/remake](https://www.npmjs.com/package/@itsjonq/remake?activeTab=readme) package to create files out of templates.
```

File: README.md
Comment (obulat) [inline README.md:133]:
```suggestion
You can also create all those files manually or by running the following commands:
```

File: package.json
Comment (obulat) [inline package.json:66]: I wonder if we could follow the example of other scripts and add a prefix to group similar commands. `create:`, maybe? So it would be `create:story`, `create:component-storybook-test`, `create:component-unit-test` etc...

**Review by obulat — COMMENTED** [review]
I really love this generator :)
I've added some suggestions, mainly spell-checking. I will test the code locally on Monday. Wish there was a way to not use `case.kebab` to convert the title case component name to kebab case so we wouldn't need to repeat the name twice.

File: README.md
Comment (obulat) [inline README.md:126]: I realized that we need `PascalCase`, and not `camelCase` for the first instance of the component name.

File: package.json
Comment (anton202) [reply package.json:66]: yeah, sure

**Comment by anton202** [comment]
@obulat thank you for the review, and for correcting my speling mistakes :)
i created a bash script that transforms the given name for the component from PascalCase to kebab-case.
so now we can use the command like this: ` pnpm run create:component VButton`

File: bin/create-component.sh
Comment (obulat) [inline bin/create-component.sh:6]:
```suggestion
componentNameInKebabCase=$(echo "$1" | sed 's/\(.\)\([A-Z]\)/\1-\2/g' | tr '[:upper:]' '[:lower:]')
```

File: bin/create-component.sh
Comment (obulat) [inline bin/create-component.sh:8]:
```suggestion
pnpm run create:component-scaffolding "$1" "$componentNameInKebabCase"
```

File: bin/create-component.sh
Comment (obulat) [inline bin/create-component.sh:5]:
```suggestion
# transforms component name from PascalCase to kebab-case
```

File: bin/add-entry-to-tsconfig.js
Comment (obulat) [inline bin/add-entry-to-tsconfig.js:5]:
```suggestion
let component_name = process.argv[2]
```

File: bin/add-entry-to-tsconfig.js
Comment (obulat) [inline bin/add-entry-to-tsconfig.js:7]:
```suggestion
if (component_name === undefined)
```

File: bin/add-entry-to-tsconfig.js
Comment (obulat) [inline bin/add-entry-to-tsconfig.js:13]:
```suggestion
  `src/components/${component_name}/${component_name}.vue`
```

File: bin/add-entry-to-tsconfig.js
Comment (obulat) [inline bin/add-entry-to-tsconfig.js:17]:
```suggestion
console.log(`added component ${component_name} to tsconfig.ts`)
```

**Review by obulat — CHANGES_REQUESTED** [review]
I've tested this locally, and now have a beautiful well-tested VButtonGreen component 🎉 

Thank you for your contributions, @anton202! I will approve the PR after the spell checking suggestions are fixed, and it can be merged after a second review from the Openverse frontend core team.

**Comment by sarayourfriend** [comment]
I've just noticed that running the command without an argument results in a rather ugly set of output:

```
➜  openverse-frontend git:(component_scafolding) pnpm create:component

> openverse-frontend@2.2.1 create:component /home/sara/projects/openverse-frontend
> ./bin/create-component.sh


> openverse-frontend@2.2.1 create:component-scaffolding /home/sara/projects/openverse-frontend
> npm-run-all "create:component-sfc --output=src/components --name={1}" "create:story --output=src/components/{1}/meta --name={1}" "create:component-unit-test --output=test/unit/specs/components/{1} --name={1} --fileName={2}" "create:component-storybook-test --output=test/storybook/visual-regression --fileName={2} --name={1}" "add-component-to-tsconfig {1}" --


> openverse-frontend@2.2.1 create:component-sfc /home/sara/projects/openverse-frontend
> remake component "--output=src/components" "--name="

Please provide a name. You can do it like:
--name=MyReactComponent or --name MyReactComponent

> openverse-frontend@2.2.1 create:story /home/sara/projects/openverse-frontend
> remake story "--output=src/components//meta" "--name="

Please provide a name. You can do it like:
--name=MyReactComponent or --name MyReactComponent

> openverse-frontend@2.2.1 create:component-unit-test /home/sara/projects/openverse-frontend
> remake component-unit-test "--output=test/unit/specs/components/" "--name=" "--fileName="

Please provide a name. You can do it like:
--name=MyReactComponent or --name MyReactComponent

> openverse-frontend@2.2.1 create:component-storybook-test /home/sara/projects/openverse-frontend
> remake component-storybook-test "--output=test/storybook/visual-regression" "--fileName=" "--name="

Please provide a name. You can do it like:
--name=MyReactComponent or --name MyReactComponent

> openverse-frontend@2.2.1 add-component-to-tsconfig /home/sara/projects/openverse-frontend
> node bin/add-entry-to-tsconfig.js

/home/sara/projects/openverse-frontend/bin/add-entry-to-tsconfig.js:8
  throw new Error('please provide component name')
  ^

Error: please provide component name
    at Object.<anonymous> (/home/sara/projects/openverse-frontend/bin/add-entry-to-tsconfig.js:8:9)
    at Module._compile (node:internal/modules/cjs/loader:1155:14)
    at Object.Module._extensions..js (node:internal/modules/cjs/loader:1209:10)
    at Module.load (node:internal/modules/cjs/loader:1033:32)
    at Function.Module._load (node:internal/modules/cjs/loader:868:12)
    at Function.executeUserEntryPoint [as runMain] (node:internal/modules/run_main:81:12)
    at node:internal/main/run_main_module:22:47
 ELIFECYCLE  Command failed with exit code 1.
ERROR: "add-component-to-tsconfig" exited with 1.
 ELIFECYCLE  Command failed with exit code 1.
 ELIFECYCLE  Command failed with exit code 1.
```

@anton202 Can you create an issue to follow up on this work to better handle incorrect input (I guess just print to stderr if `$1` is empty in `create-component.sh`?).

File: bin/add-entry-to-tsconfig.js
Comment (sarayourfriend) [inline bin/add-entry-to-tsconfig.js:15]: We use 2 spaces for indentation. Using 3 here causes the entire file to output with changes on every line.

```suggestion
fs.writeFileSync('./tsconfig.json', stringify(tsconfig, null, 2), { flag: 'w' })
```

I also noticed that doing it this way (with `comment-json`) creates other changes in the file that would need to be fixed with the linter:

```diff
diff --git a/tsconfig.json b/tsconfig.json
index b8e8fe59f..5e1c95475 100644
--- a/tsconfig.json
+++ b/tsconfig.json
@@ -6,7 +6,11 @@
     "jsx": "preserve",
     "target": "ES2018",
     "module": "ESNext",
-    "lib": ["DOM", "ESNext", "ESNext.AsyncIterable"],
+    "lib": [
+      "DOM",
+      "ESNext",
+      "ESNext.AsyncIterable"
+    ],
     "noEmit": true,
     "isolatedModules": true,
     "skipLibCheck": true,
@@ -22,11 +26,23 @@
     /* Module Resolution Options */
     "moduleResolution": "Node",
     "resolveJsonModule": true,
-    "types": ["@types/jest", "@nuxt/types", "@nuxtjs/i18n", "@pinia/nuxt"],
-    "typeRoots": ["./typings", "./node_modules/@types"],
+    "types": [
+      "@types/jest",
+      "@nuxt/types",
+      "@nuxtjs/i18n",
+      "@pinia/nuxt"
+    ],
+    "typeRoots": [
+      "./typings",
+      "./node_modules/@types"
+    ],
     "paths": {
-      "~/*": ["./src/*"],
-      "~~/*": ["./*"]
+      "~/*": [
+        "./src/*"
+      ],
+      "~~/*": [
+        "./*"
+      ]
     },
     "baseUrl": "."
   },
@@ -81,7 +97,6 @@
     "src/components/VHeader/VPageLinks.vue",
     "src/components/VHeader/VSearchBar/**.vue",
     "src/components/VImageDetails/VRelatedImages.vue",
-
     "src/pages/about.vue",
     "src/pages/feedback.vue",
     "src/pages/search-help.vue",
@@ -93,10 +108,13 @@
      * with other tools we rely on.
      */
     "src/constants/**/*",
-    "src/server-middleware/healthcheck.js"
+    "src/server-middleware/healthcheck.js",
     /** End of non TypeScript files section */
+    "src/components/VSarasComponent/VSarasComponent.vue"
+  ],
+  "exclude": [
+    "node_modules"
   ],
-  "exclude": ["node_modules"],
   "vueCompilerOptions": {
     "experimentalCompatMode": 2,
     "experimentalTemplateCompilerOptions": {
@@ -106,4 +124,4 @@
     },
     "target": 2.7
   }
-}
+}
\ No newline at end of file
```

The linter fixes and catches all of these but it causes the initial attempt at committing to fail due to the new changes:

```diff
diff --git a/tsconfig.json b/tsconfig.json
index 5e1c95475..459484e0e 100644
--- a/tsconfig.json
+++ b/tsconfig.json
@@ -6,11 +6,7 @@
     "jsx": "preserve",
     "target": "ES2018",
     "module": "ESNext",
-    "lib": [
-      "DOM",
-      "ESNext",
-      "ESNext.AsyncIterable"
-    ],
+    "lib": ["DOM", "ESNext", "ESNext.AsyncIterable"],
     "noEmit": true,
     "isolatedModules": true,
     "skipLibCheck": true,
@@ -26,23 +22,11 @@
     /* Module Resolution Options */
     "moduleResolution": "Node",
     "resolveJsonModule": true,
-    "types": [
-      "@types/jest",
-      "@nuxt/types",
-      "@nuxtjs/i18n",
-      "@pinia/nuxt"
-    ],
-    "typeRoots": [
-      "./typings",
-      "./node_modules/@types"
-    ],
+    "types": ["@types/jest", "@nuxt/types", "@nuxtjs/i18n", "@pinia/nuxt"],
+    "typeRoots": ["./typings", "./node_modules/@types"],
     "paths": {
-      "~/*": [
-        "./src/*"
-      ],
-      "~~/*": [
-        "./*"
-      ]
+      "~/*": ["./src/*"],
+      "~~/*": ["./*"]
     },
     "baseUrl": "."
   },
@@ -112,9 +96,7 @@
     /** End of non TypeScript files section */
     "src/components/VSarasComponent/VSarasComponent.vue"
   ],
-  "exclude": [
-    "node_modules"
-  ],
+  "exclude": ["node_modules"],
   "vueCompilerOptions": {
     "experimentalCompatMode": 2,
     "experimentalTemplateCompilerOptions": {
@@ -124,4 +106,4 @@
     },
     "target": 2.7
   }
-}
\ No newline at end of file
+}
```

I'm not sure what the best way to avoid this is. The only thing I can think of is to directly modify the file as a string and then write it back, without parsing and strigifying the JSON back out. If we did this, I'd probably recommend just adding the new component to the front of the list to make it easier to find the consistent insertion point.

**Review by sarayourfriend — CHANGES_REQUESTED** [review]
It looks like there are some linting changes that need to be applied to the `add-to-tsconfig.js` file. Additionally, I noticed some issues with the output of that module and with some inconsistencies in how the `comment-json` decides to write out the JSON (compared to how our lint configuration expects the file to look).

File: bin/add-entry-to-tsconfig.js
Comment (anton202) [reply bin/add-entry-to-tsconfig.js:15]: @sarayourfriend i`ll try that aproach.  regarding the create-component.sh, i`ll add some input validation in this PR

Comment (zackkrida) [reply bin/add-entry-to-tsconfig.js:15]: What if we add `eslint --fix tsconfig.json` to the end of the `create:component-scaffolding` command?

**Comment by obulat** [comment]
@anton202, I've added some changes to fix the linting failures in commit [3a917ba](https://github.com/WordPress/openverse-frontend/pull/1998/commits/3a917ba50e66c8a86af11900722e56e52d1ab9b1). I really didn't expect that the changes had to be added in the pre-commit `exclude` and `.prettierignore` at the same time.  With these changes, I think this PR is ready to go 🚀
I'll approve it when you un-draft it.

**Review by obulat — APPROVED** [review]

**Comment by krysal** [comment]
@WordPress/openverse-frontend this could benefit from another review.

File: README.md
Comment (zackkrida) [inline README.md:174]:
```suggestion
### How to create a component
```

Since the parent heading is `##`.

File: README.md
Comment (zackkrida) [inline README.md:181]:
```suggestion
```shell
# Make sure you have run `pnpm install` at least once before.
pnpm run create:component [component name in PascalCase]
for example: pnpm run create:component VButtonGreen
```
```

File: README.md
Comment (zackkrida) [inline README.md:185]:
```suggestion
tsconfig.ts `include` property. Each file will have a basic template to start
```

File: README.md
Comment (zackkrida) [inline README.md:186]:
```suggestion
working with. We use the
```

File: README.md
Comment (zackkrida) [inline README.md:190]:
```suggestion
You can also create all those files manually by running the following
```

File: bin/create-component.sh
Comment (zackkrida) [inline bin/create-component.sh:9]:
```suggestion
    # transforms component name from PascalCase to kebab-case
```

Is this meant to be a comment?

File: bin/create-component.sh
Comment (zackkrida) [inline bin/create-component.sh:15]:
```suggestion
    # if you try to commit your newly created component,
```

File: .remake/component/remake-name/remake-name.vue
Comment (zackkrida) [inline .remake/component/remake-name/remake-name.vue:1]:
```suggestion
<template>
	<div />
</template>
```

Could we populate these components with an empty `div`, so by default they pass lint checking?

File: bin/create-component.sh
Comment (anton202) [reply bin/create-component.sh:9]: oops, yep

File: .remake/component/remake-name/remake-name.vue
Comment (anton202) [reply .remake/component/remake-name/remake-name.vue:1]: fine by me

**Review by zackkrida — APPROVED** [review]
Awesome work. LGTM!

**Pull request merged**
